cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java

```diff
@@ -0,0 +1,229 @@
+   * @return returns {@link LocationIdentifier}
+   * @throws IllegalArgumentException application id is not present in the property map
+   */
+  LocationIdentifier getLocationIdentifier(Map<String, String> propertyMap) {
+
+    String namespaceId = propertyMap.get(TAG_NAMESPACE_ID);
+    String applicationId = propertyMap.get(TAG_APPLICATION_ID);
+
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(applicationId),
+                                String.format("%s is expected but not found in the context %s",
+                                              TAG_APPLICATION_ID, propertyMap));
+
+    return new LocationIdentifier(namespaceId, applicationId);
+  }
+
+  /**
+   * Returns outpustream for log file created as: <basePath>/namespaceId/applicationId/<filePath>
+   *
+   * @param locationIdentifier location identifier for this event
+   * @param filePath           filePath for this event
+   * @return returns {@link LocationOutputStream} for an event
+   * @throws IOException throws exception while creating a file
+   */
+  OutputStream getLocationOutputStream(LocationIdentifier locationIdentifier, String filePath) throws IOException {
+    if (activeLocations.containsKey(locationIdentifier)) {
+      return activeLocations.get(locationIdentifier);
+    }
+
+    Location logFile = getLogLocation(locationIdentifier).append(filePath);
+    Location logDir = getParent(logFile);
+
+    if (logDir == null) {
+      // this should never happen
+      LOG.error("Parent Directory for {} is null", logFile.toURI().toString());
+      throw new IOException(String.format("Parent Directory for %s is null", logFile.toURI().toString()));
+    }
+
+    // check if parent directories exist
+    mkdirsIfNotExists(logDir, dirPermissions);
+
+    LOG.info("Created log directory: {}", logDir.toURI().toString());
+    if (logFile.exists()) {
+      // The file name for a given application exists if the appender was stopped and then started again but file was
+      // not rolled over. In this case, since the roll over size is typically small, we can rename the old file and
+      // copy its contents to new file and delete old file.
+      long now = System.currentTimeMillis();
+
+      // rename existing file to temp file
+      Location tempLocation = logFile.renameTo(logDir.append("temp-" + Long.toString(now)));
+
+      if (tempLocation == null) {
+        throw new IOException(String.format("Can not rename file %s", logFile.toURI().toString()));
+      }
+
+      try (InputStream inputStream = tempLocation.getInputStream()) {
+        // create new file and open outputstream on it
+        logFile.createNew(filePermissions);
+        // TODO: Handle existing file in a better way rather than copying it over
+        OutputStream outputStream = new LocationOutputStream(logFile, logFile.getOutputStream());
+        activeLocations.put(locationIdentifier, (LocationOutputStream) outputStream);
+        ByteStreams.copy(inputStream, outputStream);
```

Comment:
oh, here you don't want to close the output stream, as the appends will be using this output stream right?. you can just call flush on output stream in that case, after copy.